neon-mmd/websurfx
# ✨ Compression and encryption for the cached search results

## What does this PR do?

This PR provides the compression and encryption of cached search results by providing four conditionally compiled features in rust, which looks something like this:
 
- `Default`: No compression and encryption
- `Compress-cache-results:`  Uses compression algorithm like `brotli` (which is considered good for text based compression)
- `Encrypt-cache-results:` Uses an encryption algorithm like `chacha20poly1305` (which is considered very strong encryption and is widely adopted in the `rust` programming language). Also, the generated encryption should change on each restart of the `websurfx` server, which help against attacks and improve the privacy of the user. 
-  `cec-cache-results:` Uses both compression and encryption to first compress the results then encrypt it and then compress it back again to get the advantages of both encryption and compression. 

## Why is this change important?

This change is essential as by providing this feature:
- By compressing the cached search results, the search engine would consume less amount of memory required for storing cache results, as well as improve the speed of transferring cache results over the internet. The size of the cache results become smaller after compression. So, if the cache results were to be stored on a Redis server provided by a VPS provider.
- By encrypting cache results, this can help improve user privacy by not allowing the Redis VPS provider to know what is being cached on their Redis server. Though, the cache results size become larger after encryption.
- By using both the encryption and compression with `cec-cache-results` feature, we provide the advantages of both compression and encryption by first compressing the results and the encrypting it. Then compressing it again so that the size of the cache result remains nearly the same as it was after the first compression. 

## How to test this PR locally?

It can be tested by installing and running `Websurfx` as mentioned in the `docs` and on the `readme` and by launching the browser and thoroughly testing. By using a memory profiler to check whether the app consumes less memory when the cached search results are compressed with the default non-compressed cached search results. Also, checking whether the results are getting encrypted when the `Encrypted-cache-results` features is enabled.  

## Author's checklist

- [x] provide the three conditionally compilation features
- [x] implement code for the `Compress-cache-results` feature.
- [x] implement code for the `Encrypted-cache-results` feature.
- [x] implement code for the `cec-cache-results` feature.
- [x] bump the app version to `v1.9.0`

## Related issues

Closes #416
